backend/src/index.spec.js

```diff
@@ -0,0 +1,190 @@
+const {createTestClient} = require('apollo-server-testing');
+const {ApolloServer, gql} = require('apollo-server');
+
+const typeDefs = require('./TypeDefs');
+const resolvers = require('./Resolvers');
+const PostDataStore = require('./datastore/PostDataStore');
+const UserDataStore = require('./datastore/UserDataStore');
+
+const GET_POSTS = gql`
+    query get_posts {
+        posts {
+            title
+            votes
+        }
+    }
+`;
+const GET_USERS = gql`
+    query get_user {
+        users {
+            name
+        }
+    }
+`;
+
+const MUTATE_WRITE = gql`
+    mutation($postInput: PostInput!) {
+        write(post: $postInput) {
+            id
+            title
+            votes
+            author {
+                name
+            }
+        }
+    }
+`;
+
+const MUTATE_UPVOTE = gql`
+    mutation($id: ID!, $voter: UserInput!) {
+        upvote(id: $id, voter: $voter) {
+            id
+            title
+            votes
+        }
+    }
+`;
+
+describe('server', () => {
+    let server;
+    beforeEach(() => {
+        const userDataStore = new UserDataStore([
+            {
+                name: 'Max Mustermann'
+            }
+        ]);
+        const postDataStore = new PostDataStore(userDataStore, [
+            {
+                id: 1,
+                title: 'Test Message 1',
+                votes: [],
+                author_id: 'Max Mustermann'
+            },
+            {
+                id: 2,
+                title: 'Test Message 2',
+                votes: ['Max Mustermann'],
+                author_id: 'Max Mustermann'
+            }
+        ]);
+        server = new ApolloServer({
+            typeDefs,
+            resolvers,
+            dataSources: () => ({
+                userDataStore: userDataStore,
+                postsDataStore: postDataStore
```

Comment:
In fact, if you declare local variables `userDataStore` and `postsDataStore` at the outmost scope, you can let your test re-initialize and change these datasources. Then you can share a single `const server = new ApolloServer()` for all your tests and you can DRY this up.

Only the state in your datasources change between your tests, `ApolloServer` itself is effectively stateless.